proto/ord_schema.proto

```diff
@@ -237,52 +237,69 @@ message CompoundIdentifier {
     // Protein data bank ID (for enzymes)
     PDB_ID = 13;
   }
-  IdentifierType type = 1;
-  // Custom compound identifier 
-  string type_custom = 2;
-  // Value of the compound identifier; certain types (e.g., pubchem_cid) may
+  oneof kind {
+    IdentifierType type = 1;
+    string custom = 2;
+  }
+  // Value of the compound identifier; certain types (e.g., PUBCHEM_CID) may
   // cast the string as an integer for downstream processing and validation.
   string value = 3;
 }
 
-message ReactionSetup {
-  enum VesselType {
-    VESSEL_TYPE_UNKNOWN = 0;
-    ROUND_BOTTOM_FLASK = 1;
-    VIAL = 2;  // Regular glass vial
-    WELL_PLATE = 3;
-    MICROWAVE_VIAL = 4;
-    TUBE = 5;
-    CONTINUOUS_STIRRED_TANK_REACTOR = 6;
-    PACKED_BED_REACTOR = 7;
-  }
-  VesselType vessel_type = 1;
-  // Used to describe an UNKNOWN vessel type or elaborate.
-  string vessel_type_custom = 2;
-  enum VesselMaterial {
-    VESSEL_MATERIAL_UNKNOWN = 0;
-    GLASS = 1;
-    POLYPROPYLENE = 2;
-    PLASTIC = 3;
-  }
-  VesselMaterial vessel_material = 3;
-  string vessel_material_custom = 4;
-  enum VesselPrep {
-    VESSEL_PREP_UNKNOWN = 0;
-    NONE = 1;
-    OVEN_DRIED = 2;
+message Vessel {
+  message VesselForm {
+    enum VesselFormType {
+      UNSPECIFIED = 0;
+      ROUND_BOTTOM_FLASK = 1;
+      VIAL = 2;
+      WELL_PLATE = 3;
+      MICROWAVE_VIAL = 4;
+      TUBE = 5;
+      CONTINUOUS_STIRRED_TANK_REACTOR = 6;
+      PACKED_BED_REACTOR = 7;
+    }
+    oneof kind {
+      VesselFormType type = 1;
+      string custom = 2;
+    }
   }
-  VesselPrep vessel_prep = 5;
-  // Used to describe an UNKNOWN vessel prep.
-  string vessel_prep_custom = 6;
-  // Size (volume) of the vessel.
-  Volume vessel_volume = 7;
+  message VesselMaterial {
+    enum VesselMaterialType {
+      UNSPECIFIED = 0;
+      GLASS = 1;
+      POLYPROPYLENE = 2;
+      PLASTIC = 3;
+    }
+    oneof kind {
+      VesselMaterial type = 1;
+      string custom = 2;
+    }
+  }
+  message VesselPrep {
+    enum VesselPrepType {
+      UNSPECIFIED = 0;
+      NONE = 1;
+      OVEN_DRIED = 2;
+    }
+    oneof kind {
+      VesselPrepType type = 1;
+      string custom = 2;
+    }
+  }
+  VesselForm form = 1;
```

Comment:
Thinking about the use case here -- given a ```vessel```,

```
vessel.form.type = vessel.form.VIAL
vessel.material.type = vessel.material.GLASS
vessel.prep.custom = 'pre-cooled in fridge'
```

Reply:
Having to go one level deeper into ```type``` isn't ideal. I realize that having separate (non-exclusive) ```vessel.material``` and ```vessel.material_custom``` feels less elegant, but do you think it might encourage people to use the non-custom field?

Reply:
Yeah `Vessel` is the tricky one here since it has a bunch of different enums.

I think your suggestion of `type_custom` elsewhere is probably the right way to go; that will avoid needing more nesting for these cases. I agree that we should avoid the type of stutter you highlighted above. We'll need to bite the bullet and have some awkward prefixes for the unspecified enum values, but I think that's better than extra nesting (and the style guide agrees).

Reply:
Is it worth pulling out VesselForm, VesselMaterial, and VesselPrep to be top-level messages so they can all use UNSPECIFIED?

Reply:
I'd say no; that will require the type of nesting you highlighted above and people aren't going to be using the default enum value anyway---it will only be a check in the validation script.

Reply:
(note that this might change if we decide to add a details field; I think factoring out as a separate message makes sense if you need to carry more than just type/type_custom)